### PR TITLE
[FSSDK-9079] fix: send odp event empty values

### PR DESCRIPTION
**Pull request by andrewleap-optimizely**
## Summary
- Validate SendOdpEvent empty action and type parameters
- Generate new idempotence_id for every OdpEvent

## Test plan
added tests to:
- OdpEventManagerTests.cs
- OptimizelyTest.cs

## Ticket
[FSSDK-9079](https://jira.sso.episerver.net/browse/FSSDK-9079)

### CODE REVIEW

File: OptimizelySDK/Odp/OdpEventManager.cs [hunk 1]
```diff
@@ -429,7 +431,42 @@ private bool InvalidDataFound(Dictionary<string, object> data)
...
+                if (lowerCaseKey == Constants.FS_USER_ID_ALIAS || lowerCaseKey == Constants.FS_USER_ID)
+                {
+                    identifiers.Remove(kvp.Key);
+                    identifiers[Constants.FS_USER_ID] = kvp.Value;
+                    break;
+                }
+            }
+
+            return identifiers;
         }
```
Comment (mikechu-optimizely) [inline OptimizelySDK/Odp/OdpEventManager.cs:470]:
```suggestion
        /// <summary>
        /// Convert critical identifiers to the correct format for ODP
        /// </summary>
        /// <param name="identifiers">Collection of identifiers to validate</param>
        /// <returns>New version of identifiers with corrected values</returns>
        private Dictionary<string, string> ConvertCriticalIdentifiers(Dictionary<string, string> identifiers)
        {
            if (identifiers.ContainsKey(Constants.FS_USER_ID))
            {
                return identifiers;
            }

            var newIdentifiers = new Dictionary<string, string>(identifiers.Count, StringComparer.OrdinalIgnoreCase);
            foreach (var kvp in identifiers)
            {
                if (kvp.Key.Equals(Constants.FS_USER_ID_ALIAS) || kvp.Key.Equals(Constants.FS_USER_ID))
                {
                    newIdentifiers.Add(Constants.FS_USER_ID, kvp.Value);
                }
                else
                {
                    newIdentifiers.Add(kvp.Key, kvp.Value);
                }
            }

            return newIdentifiers;
        }
```

**Review by mikechu-optimizely — CHANGES_REQUESTED** [review]
Some points to consider 👇

File: OptimizelySDK/Odp/OdpEventManager.cs (hunk 1 above)
Comment (mikechu-optimizely) [reply OptimizelySDK/Odp/OdpEventManager.cs:470]: Nope.  Probably more like 

```
...
var newIdentifiers = new Dictionary<string, string>(identifiers.Count);
            foreach (var kvp in identifiers)
            {
                if (kvp.Key.Equals(Constants.FS_USER_ID_ALIAS, StringComparison.OrdinalIgnoreCase) || 
                    kvp.Key.Equals(Constants.FS_USER_ID, StringComparison.OrdinalIgnoreCase))
                {
...
```

File: OptimizelySDK.Tests/OptimizelyTest.cs [hunk 2]
```diff
@@ -6166,5 +6169,64 @@ public static void SetCulture(string culture)
...
         #endregion Test Culture
+
+        #region Test SendOdpEvent
+
+        [Test]
+        public void TestSendOdpEventNullAction()
+        {
+            var optly = new Optimizely(TestData.OdpIntegrationDatafile, logger: LoggerMock.Object, odpManager: OdpManagerMock.Object);
+            optly.SendOdpEvent("type", null, identifiers: new Dictionary<string, string>(), data: new Dictionary<string, object>());
+            LoggerMock.Verify(l => l.Log(LogLevel.ERROR, "ODP action is not valid (cannot be empty)."),
```
[the remark this answers is not in the document]
Comment (mikechu-optimizely) [reply OptimizelySDK.Tests/OptimizelyTest.cs:6180]: I'd vote using the Constant since, you know, @jaeopt  might change the messaging on us 😉 just kidding

**Review by mikechu-optimizely — APPROVED** [review]
LGTM. Thanks 🙏

File: OptimizelySDK/Optimizely.cs [hunk 3]
```diff
@@ -1362,6 +1362,17 @@ internal void IdentifyUser(string userId)
             Dictionary<string, object> data
```
Comment (jaeopt) [inline OptimizelySDK/Optimizely.cs:1362]: Don't we need to assign default values for `type = null` and `data = [:]`?

File: OptimizelySDK/Optimizely.cs [hunk 4]
```diff
@@ -1362,6 +1362,17 @@ internal void IdentifyUser(string userId)
...
+            {
+                Logger.Log(LogLevel.ERROR, Constants.ODP_INVALID_ACTION_MESSAGE);
+                return;
+            }
+
+            if (String.IsNullOrEmpty(type))
+            {
+                type = Constants.ODP_EVENT_TYPE;
+            }
+
```
Comment (jaeopt) [inline OptimizelySDK/Optimizely.cs:1375]: Where do we check if `identifiers` has non-empty?

File: OptimizelySDK/Optimizely.cs (hunk 3 above)
Comment (jaeopt) [reply OptimizelySDK/Optimizely.cs:1362]: If we add "null" type default, can we remove "(typically `fullstack` from server-side SDK events)" in line 1357.

File: OptimizelySDK/Utils/CollectionExtensions.cs [hunk 5]
```diff
@@ -43,16 +42,14 @@ public static class CollectionExtensions
 
             if (right == null)
             {
-                return left;
+                return;
             }
 
             foreach (var kvp in right.Where(
                          kvp => !left.ContainsKey(kvp.Key)))
             {
```
Comment (jaeopt) [inline OptimizelySDK/Utils/CollectionExtensions.cs:50]: We can make it more clear with a comment about "left has higher priority when keys conflict".

**Review by jaeopt — CHANGES_REQUESTED** [review]
Looks good! A couple of changes suggested.

File: OptimizelySDK/Optimizely.cs (hunk 4 above)
Comment (andrewleap-optimizely) [reply OptimizelySDK/Optimizely.cs:1375]: Looks like the check is in OdpEventManager.SendEvent

File: OptimizelySDK/Optimizely.cs (hunk 3 above)
Comment (andrewleap-optimizely) [reply OptimizelySDK/Optimizely.cs]: I've made this change. (If it looks good, I'll update the csharp-testapp.)

**Review by jaeopt — APPROVED** [review]
LGTM

**Review by mikechu-optimizely — APPROVED** [review]
LGTM

**Pull request merged**